friendshipper/src-tauri/src/auth/router.rs

```diff
@@ -1,25 +1,29 @@
-use std::sync::Arc;
-
 use anyhow::{anyhow, Result};
 use axum::extract::State;
 use axum::routing::{get, post};
-use axum::{debug_handler, Json, Router};
+use axum::{Json, Router};
 use ethos_core::AWSClient;
 
+use crate::engine::EngineProvider;
 use crate::APP_NAME;
 use ethos_core::clients::aws::ensure_aws_client;
 use ethos_core::types::errors::CoreError;
 
 use crate::state::AppState;
 
-pub fn router(shared_state: Arc<AppState>) -> Router {
+pub fn router<T>() -> Router<AppState<T>>
+where
+    T: EngineProvider,
```

Comment:
Just commenting on the first instance of this.

In reality we should only need the type args on handlers that need to make method args. Axum does allow for extracting substates, so we could have something like:

```rust
pub type AppState<T> {
  pub config: ConfigState,
  pub engine: Arc<T>
}
```

and then most of the handlers can be like:

```rust
async fn foo(State(config): State<ConfigState>) -> Result;
```

while others can be
```rust
async fn foo_with_engine<T>(State(config): State<ConfigState>, State(engine): State<T>) -> Result;
```

Axum will help with these extractions, but in order to get these substates we need to create routers that group handlers by which states they require. 

That said, I don't have a _huge_ problem with putting these type constraints everywhere, but it is tedious especially when you don't actually need any engine methods.

Reply:
I'm also trying to imagine how this pattern will expand when we have storage provider traits and kubernetes client provider traits, etc.